packages: Refactor setup scripts

erick0z opened the pull request:
this series of commits refactor the setup script allowing new features,
needed for some use cases and adding a better user experience.

- Splitted common functions to scripts/pkglib.sh

Fixes #166

Commit:
runtime: Update setup scripts

    this commit refactor the setup script allowing new features,
    needed for some use cases and adding a better user experience.

    Fixes #166

Signed-off-by: Erick Cardona <erick.cardona.ruiz@intel.com>

Commit:
shim: Update setup scripts

    this commit refactor the setup script allowing new features,
    needed for some use cases and adding a better user experience.

    Fixes #166

Signed-off-by: Erick Cardona <erick.cardona.ruiz@intel.com>

Commit:
proxy: Update setup scripts

this commit refactor the setup script allowing new features,
needed for some use cases and adding a better user experience.

Fixes #166

Signed-off-by: Erick Cardona <erick.cardona.ruiz@intel.com>

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -6,41 +6,106 @@
+}
 
 function changelog_update {
     d=$(date +"%a, %d %b %Y %H:%M:%S %z")
```
Replace this please.

`d=$(date -R)`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -6,41 +6,106 @@
+    # the script will run flawlessly.
+
+    # Make sure this script is called from ./
+    [ "$SCRIPT_DIR" != "." ] && echo "The script must be called from its base dir." && exit 1 || :
```
why is this needed?

gorozco1 replied:
For clarity better to use: 

`[ "$SCRIPT_DIR" != "." ] && echo "The script must be called from its base dir." && exit 1 || true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
-    [ -f $TMPDIR/debian.series ] && rm $TMPDIR/debian.series || :
+        $OBS_WORKDIR
+    rm $OBS_WORKDIR/*.patch
+    [ -f $OBS_WORKDIR/debian.series ] && rm $OBS_WORKDIR/debian.series || :
```
`[ -f $OBS_WORKDIR/debian.series ] && rm $OBS_WORKDIR/debian.series || true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
 
-    [ -f debian.series ] && cp debian.series $TMPDIR || :
-    cd $TMPDIR
+    [ -f debian.series ] && cp debian.series $OBS_WORKDIR || :
```
`[ -f debian.series ] && cp debian.series $OBS_WORKDIR || true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
+
+function local_build()
+{
+    [ ! -e /packages ] && mkdir /packages || :
```
`|| true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
+function local_build()
+{
+    [ ! -e /packages ] && mkdir /packages || :
+    [ ! -e /logs ] && mkdir /logs || :
```
`|| true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
+
+    pushd $OBS_WORKDIR
+
+    BUILD_ARGS=('--local-package' '--no-verify' '--noservice' '--trust-all-projects' '--keep-pkgs=/packages')
```
`--no-verify` ?

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
+function main()
+{
+
+    [ -n "$VERBOSE" ] && set -x || :
```
`|| true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
+    VERSION=$cc_runtime_version
+    RELEASE=$(cat release)
+    OBS_RUNTIME_REPO=${OBS_RUNTIME_REPO:-home:clearcontainers:clear-containers-3-staging/cc-runtime}
+    [ -n "$APIURL" ] && APIURL="-A ${APIURL}" || :
```
`|| true`

gorozco1 commented on runtime/update_runtime.sh:
```diff
@@ -55,83 +120,177 @@ function changelog_update {
+    echo "Working directory: ${OBS_WORKDIR}"
+}
+
+OPTS=`getopt -o abclprwvCVh: --long api-url,branch,commit-id,local-build,push,obs-repository,workdir,verbose,clean,verify,help: -n 'test' -- "$@"`
```
change `code` for $(code)

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+  cat <<-EOL 
+  $SCRIPT_NAME
+
+  This script is intended to create Clear Containers 3.0 packages for the OBS 
```
I'd change this to "Clear Containers 3.x" to make it more future-proof.

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+  Options:
+
+    -l         --local-build     Build the runtime locally
+    -w	       --workdir	 Repository directory
```
Indentation.

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+    -V         --verify          Verify the environment
+    -h         --help            Display this help message
+
+    Usage example
```
s/Usage example/Usage examples:/

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+    # the script will run flawlessly.
+
+    # Make sure this script is called from ./
+    [ "$SCRIPT_DIR" != "." ] && echo "The script must be called from its base dir." && exit 1 || :
```
It might be worth adding a `die` function so you can simplify this slightly:
```
die()
{
  msg="$*"
  echo >&2 "ERROR: $msg"
  exit 1
}

  :
  :

[ "$SCRIPT_DIR" != "." ] && die "The script must be called from its base dir" || :

```

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+{
+    # This function clean generated files
+    local files=(cc-runtime.spec cc-runtime.dsc debian.control debian.rules _service)
+    for file in ${files[@]}
```
Nit: You might want a `local file` here as you have defined a local for `files`?

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+
+function local_build()
+{
+    [ ! -e /packages ] && mkdir /packages || :
```
I'd consider creating variables for paths like `/packages` and `/logs` to make maintaining these scripts easier.

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+        then
+	    echo "Perform a local build for ${distro}"
+	    osc build ${BUILD_ARGS[@]} \
+                ${distro} x86_64 *.spec | tee /logs/proxy_${distro}_build.log
```
Again, it wouldn't hurt to define a variable for `x86_64` to avoid the repetition here.

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+    [ -n "$APIURL" ] && APIURL="-A ${APIURL}" || :
+    
+    verify
+    [[ $? != 0 ]] && exit 1 || echo "Verify succeed."
```
The `verify` function exits on error, so this could be simplified to the following I think:

```
verify
echo "Verify succeeded."
```

jodh-intel commented on runtime/update_runtime.sh:
```diff
@@ -6,41 +6,106 @@
+  Options:
+
+    -l         --local-build     Build the runtime locally
+    -w	       --workdir	 Repository directory
```
Indentation oddity again.

jodh-intel commented on proxy/update_proxy.sh:
```diff
@@ -5,112 +5,269 @@
+
+function set_versions()
+{
+    source ../versions.txt
```
I'd perform the `source` at the top of the script to avoid having to call it from (various) functions.

jodh-intel commented:
This PR looks like it's going to be very useful. However, these `update_*.sh` scripts are almost identical. I suggest refactoring a little further so that the 3 scripts can be "stubs" and call functions in a `lib.sh` or something. The stubs could then:

- `source lib.sh`
- Parse their arguments
- Make calls like this:
  ```
  clean $runtime_files

  set_versions "$cc_shim_version" "$cc_shim_hash"
  
  checkout_repo "$OBS_PROXY_REPO" ...
  ```

Commit:
image: Update setup scripts

this commit refactor the setup script allowing new features,
needed for some use cases and adding a better user experience.

Fixes #166

Signed-off-by: Erick Cardona <erick.cardona.ruiz@intel.com>

Commit:
qemu-lite: Update setup scripts

this commit refactor the setup script allowing new features,
needed for some use cases and adding a better user experience.

Fixes #166

Signed-off-by: Erick Cardona <erick.cardona.ruiz@intel.com>

Commit:
kernel: Update setup scripts

this commit refactor the setup script allowing new features,
needed for some use cases and adding a better user experience.

Fixes #166

Signed-off-by: Erick Cardona <erick.cardona.ruiz@intel.com>

erick0z commented:
@jodh-intel, As you suggested, I've split some functions into a new ``scripts/pkglib.sh``. The remaining bits in each setup script are just properties needed for each component (like project names and static files, etc). Can you review this PR?

gorozco1 commented:
@erick0z this commit has the "do not merge" label is still the case?
